Index: orocos_kdl/src/chainjnttojacsolver.hpp
===================================================================
--- orocos_kdl/src/chainjnttojacsolver.hpp	(revision 34658)
+++ orocos_kdl/src/chainjnttojacsolver.hpp	(working copy)
@@ -53,15 +53,19 @@
          *
          * @return always returns 0
          */
-        int JntToJac(const JntArray& q_in,Jacobian& jac);
+      int JntToJac(const JntArray& q_in,Jacobian& jac, bool coupled=false);
         
         int setLockedJoints(const std::vector<bool> locked_joints);
+
+        std::vector<bool> getLockedJoints();
+
     private:
         const Chain chain;
         Twist t_tmp;
         Frame T_tmp;
         std::vector<bool> locked_joints_;
         int nr_of_unlocked_joints_;
+        Jacobian jac_tmp;
     };
 }
 #endif

